### PR TITLE
feat: implement training pynative model recipe

### PR DESCRIPTION
**Pull request by shekhar-rudder**
## Description of the change

- Added another implementation of the WhtService interface called PyNativeWht. For most of the methods, it is using PythonWht to implement them. The plan is to phase it out incrementally
- Added a new argument `ml_core_path` in the train function since its value is going to be different for Python and PyNative models
- Created a method `get_model_creation_ts` in pythonWht so that it can be used by PyNativeWht
- Removed `_` from `_split_material_name` in pythonWht so that it can be used by PyNativeWht

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/predictions/rudderstack_predictions/wht/pyNativeWHT.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        self.whtMaterial = whtMaterial
+
+    def init(
+        self,
+        connector: Connector,
+        session,
+        site_config_path: str,
+        project_folder_path: str,
+    ) -> None:
+        self.pythonWHT.init(connector, session, site_config_path, project_folder_path)
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/wht/pyNativeWHT.py:21]: From now on, will pythonWHT references be made only in this file?

File: src/predictions/rudderstack_predictions/processors/ProcessorFactory.py [hunk 2]
```diff
@@ -5,17 +5,16 @@
...
 class ProcessorFactory:
-    def create(mode: str, trainer, connector, session):
+    def create(mode: str, trainer, connector, session, ml_core_path: str):
         processor = None
         if mode == constants.RUDDERSTACK_MODE:
             # Lazy load K8sProcessor since kubernetes might not be installed in all environments
             from .K8sProcessor import K8sProcessor
 
-            processor = K8sProcessor
+            return K8sProcessor(trainer, connector, session, ml_core_path)
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/processors/ProcessorFactory.py:14]: Why not keep it as earlier? Just add the ml_core_path in the single return statement?

File: src/predictions/rudderstack_predictions/py_native/training.py [hunk 3]
```diff
@@ -2,36 +2,113 @@
...
+            "ml_config": {
+                "type": "object",
+                "additionalProperties": False,
+                "properties": {
+                    "data": {
+                        "type": "object",
+                        "additionalProperties": True,
+                        "properties": {
+                            "label_column": {"type": "string"},
+                            "prediction_horizon_days": {"type": "integer"},
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/py_native/training.py:32]: These are the mandatory fields but there are many other optional params. They will be added later right?

File: src/predictions/rudderstack_predictions/py_native/training.py [hunk 4]
```diff
@@ -2,36 +2,113 @@
...
-        pass
+        import os
+        import yaml
+
+        homedir = os.path.expanduser("~")
+        with open(os.path.join(homedir, ".pb/siteconfig.yaml"), "r") as f:
+            creds = yaml.safe_load(f)["connections"]["test"]["outputs"]["env"]
+        input_model_refs = self.build_spec.get("inputs", [])
+        output_filename = self._get_train_output_filepath(this)
+        site_config_path = os.getenv("SITE_CONFIG_PATH")
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/py_native/training.py:95]: How and where are the env variables set?
Also site_config_path is used above in line 91. shouldn't that come after this line and use a single site config path?

File: src/predictions/rudderstack_predictions/trainers/MLTrainer.py [hunk 5]
```diff
@@ -315,9 +315,11 @@ def check_and_generate_more_materials(
         materials: List[TrainTablesInfo],
         input_models: str,
         whtService: PythonWHT,
+        connector: Connector,
+        session,
     ):
         met_data_requirement = self.check_min_data_requirement(
-            whtService.connector, whtService.session, materials
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/trainers/MLTrainer.py:320]: why is this change required? PythonWHT still has connector and session objects right

File: src/predictions/rudderstack_predictions/wht/pyNativeWHT.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+    def get_latest_entity_var_table(self, entity_key: str) -> Tuple[str, str, str]:
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/wht/pyNativeWHT.py:23]: Can we add unit tests for these pynative functions? May be by mocking the whtmaterial (can we use the mockpb for that?)

**Review by dpatchigolla — APPROVED** [review]
Looks good. Have a few suggestions/clarifications.

File: src/predictions/rudderstack_predictions/wht/pyNativeWHT.py (hunk 1 above)
Comment (shekhar-rudder) [reply src/predictions/rudderstack_predictions/wht/pyNativeWHT.py:21]: No, the outermost train.py file will still reference pythonWHT. This is just a way of reusing the existing code in pyNativeWHT

File: src/predictions/rudderstack_predictions/py_native/training.py (hunk 3 above)
Comment (shekhar-rudder) [reply src/predictions/rudderstack_predictions/py_native/training.py:32]: Yes there are many other optional params. I don't have an opinion right now if we should add all of them (because they are too many) but yes we can add the important ones later

File: src/predictions/rudderstack_predictions/trainers/MLTrainer.py (hunk 5 above)
Comment (shekhar-rudder) [reply src/predictions/rudderstack_predictions/trainers/MLTrainer.py:320]: Yes PythonWHT still has connector and session objects but in the py native flow, MLTrainer will get the pyNativeWHT object. So the current code won't work (it should be whtService.pythonWHT.connector). So decoupled connector and session obects from whtService

File: src/predictions/rudderstack_predictions/processors/ProcessorFactory.py (hunk 2 above)
Comment (shekhar-rudder) [reply src/predictions/rudderstack_predictions/processors/ProcessorFactory.py]: Reverted back
